### ISSUE TITLE
[C++] Add concrete ArraySpan classes

### ISSUE THREAD

**Issue by pitrou** [apache/arrow#37153]
### Describe the enhancement requested

`ArraySpan` is a lighter-weight facility that `ArrayData` and `Array`, but it does not expose as many helpers, making it uneasy to use efficiently inside compute kernels.

Perhaps we should add concrete classes providing additional behaviour around `ArraySpan`, for example:
```c++

class ConcreteArraySpan {
 protected:
  ArraySpan* span;
 public:
  ConcreteArraySpan(const ArraySpan& span) : span(&span) {}
};

template <typename TypeClass>
class NumericArraySpan : public ConcreteArraySpan {
  ArraySpan* span;
 public:
  using value_type = typename TypeClass::c_type;
  using ConcreteArraySpan::ConcreteArraySpan;

  bool IsValid(int64_t index) const { ... }
  value_type GetView(int64_t index) const  { ... }
};

class BinaryArraySpan {
  ...
  bool IsValid(int64_t index) const { ... }
  std::string_view GetView(int64_t index) const { ... }
}
```

### Component(s)

C++

**Comment by pitrou** [comment]
cc @js8544 @bkietz

**Comment by felipecrv** [comment]
My concern is that the inheritance chain and type-based dispatching introduce another layer of indirection in compute kernel code that hinders readability.

Considering that kernel code is directly concerned with the dispatching of function calls based on types, time units, validity bitmaps, using single functions that can handle multiple types, adding another layer of function call dispatching would make the code harder to reason about IMO.

I'm not saying the kernel code is easy as it is now, but I would rather improve it in the dispatching paradigm it has started with than tilt it towards the type-based call dispatching paradigm (one function per type).

**Comment by pitrou** [comment]
We are not talking about virtual methods or type-based dispatching here, just concrete helper classes to avoid repeating the same low-level access code.

**Comment by felipecrv** [comment]
> We are not talking about virtual methods or type-based dispatching here, just concrete helper classes to avoid repeating the same low-level access code.

Ok, that makes sense. I was confused by the fact that the code snippet has a super class. If virtual dispatching isn't wanted we could probably drop the super-class altogether and repeat the constructor that takes an `ArraySpan` in the concrete classes containing operations.

**Comment by felipecrv** [comment]
Member functions that would fit within this superclass (like `IsNull`/`IsValid`) can be defined in `ArraySpan` itself.

**Comment by pitrou** [comment]
Also, for more context, I opened this when reading the repeated implementations in https://github.com/apache/arrow/pull/36891/files

**Comment by js8544** [comment]
I strongly agree that it is painful to operate with `ArraySpan` with such utilities missing. I'll add one and then use it in #36891.

**Comment by thisisnic** [comment]
This issue has been marked as stale because it has had no activity in the past 2 years. Please remove the stale label or comment below, or this issue will be closed in 14 days.
